api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+	var result []*ResourceProvider
+
+	listContainerInstancesInput := &ecs.ListContainerInstancesInput{}
+	listContainerInstancesInput.SetCluster(clusterName)
+	listContainerInstancesInput.SetStatus(ecs.ContainerInstanceStatusActive)
+
+	containerInstanceARNs := []*string{}
+	listContainerInstancesPagesFN := func(output *ecs.ListContainerInstancesOutput, lastPage bool) bool {
+		containerInstanceARNs = append(containerInstanceARNs, output.ContainerInstanceArns...)
+
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListContainerInstancesPages(listContainerInstancesInput, listContainerInstancesPagesFN); err != nil {
+		return nil, err
+	}
+
+	if len(containerInstanceARNs) == 0 {
+		return result, nil
+	}
+
+	describeContainerInstancesInput := &ecs.DescribeContainerInstancesInput{}
+	describeContainerInstancesInput.SetCluster(clusterName)
+	describeContainerInstancesInput.SetContainerInstances(containerInstanceARNs)
+
+	output, err := e.Client.ECS.DescribeContainerInstances(describeContainerInstancesInput)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(output.ContainerInstances) == 0 {
+		return result, nil
+	}
+
+	for _, instance := range output.ContainerInstances {
+		// it's non-intuitive, but the ports being used by the tasks live in
+		// instance.RemainingResources, not instance.RegisteredResources
+		var (
+			usedPorts       []int
+			availableCPU    bytesize.Bytesize
+			availableMemory bytesize.Bytesize
+		)
+
+		for _, resource := range instance.RemainingResources {
+			switch aws.StringValue(resource.Name) {
+			case "CPU":
+				val := aws.Int64Value(resource.IntegerValue)
+				availableCPU = bytesize.MiB * bytesize.Bytesize(val)
+
+			case "MEMORY":
+				val := aws.Int64Value(resource.IntegerValue)
+				availableMemory = bytesize.MiB * bytesize.Bytesize(val)
+
+			case "PORTS":
+				for _, port := range resource.StringSetValue {
+					port, err := strconv.Atoi(aws.StringValue(port))
+					if err != nil {
+						log.Printf("[WARN] [EnvironmentScaler] Instance %s: Failed to convert port to int: %v", aws.StringValue(instance.Ec2InstanceId), err)
+						continue
+					}
+
+					usedPorts = append(usedPorts, port)
+				}
+			}
+		}
+
+		inUse := aws.Int64Value(instance.PendingTasksCount)+aws.Int64Value(instance.RunningTasksCount) > 0
+
+		r := &ResourceProvider{
+			AgentIsConnected: aws.BoolValue(instance.AgentConnected),
+			AvailableCPU:     availableCPU,
+			AvailableMemory:  availableMemory,
+			ID:               aws.StringValue(instance.Ec2InstanceId),
+			InUse:            inUse,
+			Status:           aws.StringValue(instance.Status),
+			UsedPorts:        usedPorts,
+		}
+
+		result = append(result, r)
+	}
+
+	return result, nil
+}
+
+func (e *EnvironmentScaler) scaleDown(clusterName string, desiredScale int, asg *autoscaling.Group, unusedProviders []*ResourceProvider) error {
+	minCapacity := int(aws.Int64Value(asg.MinSize))
+	if desiredScale < minCapacity {
```

Comment:
All of this logic can be made way more simple. 
```go
func ScaleTo(environmentID string, desiredScale int, unusedProviders) error {
    // readASG is a common helper
   asg, err := readASG(fqEnvironmentID)
   
    // I believe updateASG is a common helper, or it can be moved from update_environment.go into one
    if _, err := updateASG(fqEnvironment, asg.Min, desired, asg.Max); err != nil {
        if awsErr ok := err.(awserr.Error); ok && awsErr.Code == "BelowMin || awsErr.Code == "AboveMax" {
            log.Printf("[INFO] Tried to go above/below max/min scale")
        } else {
            return err
        }
    }

    if desiredScale < asg.CurrentCapacity {
        // delete unused providers
    }

    return nil
}
```

Reply:
(this is psuedo-code of course, I don't know the literal error codes returned)

Reply:
@zpatrick It seems like the error codes returned from `UpdateAutoScalingGroup` don't actually tell us if there was a problem with going over or under the max/min...? At least, that's what I'm gathering from here: https://github.com/aws/aws-sdk-go/blob/v1.12.19/service/autoscaling/api.go#L5027

Reply:
@zpatrick With documentation not being particularly helpful, I went and generated the error IRL:
```
2017/12/14 23:04:49 [ERROR] [Worker] [1]: Failed to run job 1513321488582967461: ValidationError: Desired capacity:1 must be between the specified min size:2 and max size:2
        status code: 400, request id: f3102faf-e165-11e7-aa89-ad3b5d4b7a06
```

From what I can tell, there's not an error code/type that's specific to trying to scale beyond the min/max bounds. `ValidationError` can be returned from other places, but it's unclear to me whether scaling-beyond-bounds is the _only_ way to generate it from `autoscalingapi.UpdateAutoScalingGroup()`. I can definitely check `err.Code() == "ValidationError"`, which might be good enough; I cal also try to do some string matching within `err.Message()` if that sounds worthwhile.